pkg/frontend/quota_validation.go

```diff
@@ -0,0 +1,103 @@
+package frontend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/api/validate"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+)
+
+type QuotaValidator interface {
+	ValidateQuota(ctx context.Context, azEnv *azureclient.AROEnvironment, environment env.Interface, subscriptionID, tenantID string, oc *api.OpenShiftCluster) error
+}
+
+type quotaValidator struct{}
+
+func addRequiredResources(requiredResources map[string]int, vmSize api.VMSize, count int) error {
+	vm, ok := validate.VMSizeFromName(vmSize)
+	if !ok {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "", "The provided VM SKU %s is not supported.", vmSize)
+	}
+
+	requiredResources["virtualMachines"] += count
+	requiredResources["PremiumDiskCount"] += count
+
+	requiredResources[vm.Family] += vm.CoreCount * count
+	requiredResources["cores"] += vm.CoreCount * count
+	return nil
+}
+
+// ValidateQuota checks usage quotas vs. resources required by cluster before cluster
+// creation
+// It is a method on struct so we can make use of interfaces.
+func (q quotaValidator) ValidateQuota(ctx context.Context, azEnv *azureclient.AROEnvironment, environment env.Interface, subscriptionID, tenantID string, oc *api.OpenShiftCluster) error {
+	fpAuthorizer, err := environment.FPAuthorizer(tenantID, environment.Environment().ResourceManagerEndpoint)
+	if err != nil {
+		return err
+	}
```

Comment:
I think this is the place where we might hit AAD propagation delay.

https://github.com/Azure/ARO-RP/blob/07a08c6b59c47694b80d5168226977d4bced1acd/pkg/util/refreshable/refreshable.go#L27-L55